### ISSUE TITLE
Do not include `.git/index` when the `main` field in package.json is set to `index`

### ISSUE THREAD

**Issue by Raynos** [npm/npm-packlist#35]
When the `"main"` field in package.json is set to `index` instead of `index.js` it will attempt to include all files called `index` in the tarball when running `npm pack`

This includes `.git/index` which includes the `.git` directory which makes `npm` fail with `EISGIT` at installation time ( not publish time ).

We do not want to publish `.git` because that poisons the tarball.

Steps to reproduce :

 - edit package.json `"main"` to `index` instead of `index.js` ( can be in `npm-packlist` )
 - run `npm pack --dry-run`
 - verify that `.git/index` is printed on STDOUT

**Comment by JLLeitschuh** [comment]
This should be considered a security vulnerability.

Exposing the `.git/config` file has security ramifications as often this file holds sensitive credentials!

 - [Git ransom campaign incident report](https://about.gitlab.com/2019/05/14/git-ransom-campaign-incident-report-atlassian-bitbucket-github-gitlab/)
 - [PSA Hide Your Git Config Directory](https://laravel-news.com/psa-hide-your-gitconfig-directory)
 - [Devs, please stop serving .git to the outside world!](https://sweetness.hmmz.org/2013-06-10-devs-please-stop-serving-git-to-the-outside.html)

CC: @evilpacket

**Comment by feross** [comment]
Other files like `.DS_Store` are also getting published. Might be related to this issue.

**Comment by DanielRuf** [comment]
> Exposing the .git/config file has security ramifications as often this file holds sensitive credentials!

Not sure but I see no relevance to the reported issue which just includes `.git/index` or any file called `index`.

@feross see https://github.com/npm/npm-packlist/commit/7fcd045b7a46ef06651763e41d6c65393d509ef5 and https://github.com/npm/npm-packlist/commit/68b7c96982c2f4cf5b9d7070f07a478364bd034b

**Comment by isaacs** [comment]
As of 6.10.0, the root `.git` folder is no longer even being walked.  I think this is probably not an issue any more.  Care to check in the latest relese?

**Comment by JLLeitschuh** [comment]
@isaacs Does this need a CVE number?

**Comment by DanielRuf** [comment]
@JJLeitschuh I don't think so or can you provide a PoC which adds the mentioned file by default?

https://github.com/npm/npm-packlist/issues/35#issuecomment-507019194

> As of 6.10.0, the root .git folder is no longer even being walked.

walk !== include, it just included the .git directory for searching the files to be packed but does not pack the full `.git` directory afaik.

**Comment by isaacs** [comment]
> walk !== include

In the case of this module, it does, because it only can  include files from folders it walks, and it only includes folders containing included files.

68b7c96 fixes this issue.

PoC setup:

```
$ mkdir gh-35

$ cd gh-35

$ git init
Initialized empty Git repository in /Users/isaacs/dev/npm/npm-packlist/gh-35/.git/

$ echo '{"main":"index", "name":"testing-gh-35","version":"1.2.3"}' > package.json

$ git add package.json

$ git commit -m 'package.json'
[master (root-commit) 5a63f8c] package.json
 1 file changed, 1 insertion(+)
 create mode 100644 package.json
```

Prior to 68b7c96:

```
$ node -p 'require("../").sync()'
[ 'package.json', '.git/index' ]
```

On 68b7c96 and later:

```
$ node -p 'require("../").sync()'
[ 'package.json' ]
```

This change shipped in 8d85a56 on version 1.4.2.

Version 1.4.4 was added to the npm cli on https://github.com/npm/cli/commit/ec6236210, which was shipped in the npm cli version 6.10.0 on 2019-07-03: https://github.com/npm/cli/commit/c1522be24  https://github.com/npm/cli/releases/tag/v6.10.0

If you wish to open a CVE, that's fine, but it should note the facts: this issue is resolved and a fix has already been shipped.

**Comment by JLLeitschuh** [comment]
I will file a CVE that the `.git/index` file is included.
This is simply an information disclosure vulnerability correct? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DanielRuf** [comment]
Would be new for me that there is anything confidential in `.git/index`.

**Comment by DanielRuf** [comment]
> this issue is resolved and a fix has already been shipped.

Exactly.

**Comment by DanielRuf** [comment]
> In the case of this module, it does

But not all files like `.git/config`. This is what I mean. It does not include the complete `.git` directory including objects, hooks, ...

**Comment by JLLeitschuh** [comment]
Upon further reflection, I don't think this is a security issue. Cheers everyone, thanks for your feedback!

**Comment by michaelsbradleyjr** [comment]
I experienced this today w/ npm `6.10.1` on macOS.

See: https://npm.community/t/installing-with-git-url-results-in-spurious-git-directory-within-node-modules/8815

**Comment by michaelsbradleyjr** [comment]
Okay, I'm actually coming back with a good report. 😄 

In a fresh docker run of `node:10.16.0` I upgraded to `npm@6.10.1` and then when installing from the `git://` URL as describe in the bug report linked above, I didn't experience the bugged behavior.

So then I went back to my macOS environment with 6.10.1, ran `npm cache clear --force` and retried the install with the `git://` URL — no spurious `.git/` folder! 🎉 

Maybe node 6.10.2 could include some bug-fix logic whereby on first-run it nukes from the cache any packages that were installed from `git://` URLs?
